// ionic/dist/commands/generate.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const chalk = require("chalk");
const cli_utils_1 = require("@ionic/cli-utils");
let GenerateCommand = class GenerateCommand extends cli_utils_1.Command {
    preRun(inputs, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const project = yield this.env.project.load();
            if (project.type !== 'ionic-angular') {
                throw this.exit('Generators are only supported in Ionic Angular projects.');
            }
            const TYPE_CHOICES = ['component', 'directive', 'page', 'pipe', 'provider', 'tabs'];
            const hooks = this.env.hooks.getRegistered('command:generate');
            if (hooks.length === 0) {
                const plugin = yield cli_utils_1.promptToInstallProjectPlugin(this.env, {
                    message: `To use generators, you need to install ${chalk.green('@ionic/cli-plugin-ionic-angular')}. Install and continue?`,
                });
                if (plugin) {
                    cli_utils_1.installPlugin(this.env, plugin);
                }
                else {
                    return 1;
                }
            }
            if (!inputs[0]) {
                const response = yield this.env.prompt({
                    type: 'list',
                    name: 'type',
                    message: 'What would you like to generate:',
                    choices: TYPE_CHOICES,
                });
                inputs[0] = response['type'];
            }
            if (!TYPE_CHOICES.includes(inputs[0])) {
                throw this.exit(`${chalk.bold(inputs[0])} is not a valid generator type.`);
            }
            if (!inputs[1]) {
                const response = yield this.env.prompt({
                    name: 'name',
                    message: 'What should the name be?',
                });
                inputs[1] = response['name'];
            }
        });
    }
    run(inputs, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const [type, name] = inputs;
            yield this.env.hooks.fire('command:generate', { env: this.env, inputs, options });
            this.env.log.ok(`Generated a ${chalk.bold(type)}${type === 'tabs' ? ' page' : ''} named ${chalk.bold(name)}!`);
        });
    }
};
GenerateCommand = __decorate([
    cli_utils_1.CommandMetadata({
        name: 'generate',
        type: 'project',
        description: `Generate pipes, components, pages, directives, providers, and tabs ${chalk.bold(`(ionic-angular >= 3.0.0)`)}`,
        exampleCommands: ['', 'component', 'directive', 'page', 'pipe', 'provider', 'tabs', 'component foo', 'page Login', 'pipe MyFilterPipe'],
        inputs: [
            {
                name: 'type',
                description: `The type of generator (e.g. ${['page', 'component', 'tabs'].map(t => chalk.green(t)).join(', ')})`,
            },
            {
                name: 'name',
                description: 'The name of the component being generated',
            }
        ],
    })
], GenerateCommand);
exports.GenerateCommand = GenerateCommand;
